test/services/database/post_upvote_repository_test.dart

```diff
@@ -0,0 +1,292 @@
+
+  setUp(() async {
+    firestore = FakeFirebaseFirestore();
+    postRepository = PostRepositoryService(firestore: firestore);
+    postUpvoteRepository = PostUpvoteRepositoryService(firestore: firestore);
+  });
+
+  Future<void> assertPostUpvoteState(
+    UserIdFirestore userId,
+    PostIdFirestore postId,
+    UpvoteState expectedState, {
+    testScore = true,
+  }) async {
+    final upvoteState = await postUpvoteRepository.getUpvoteState(
+      userId,
+      postId,
+    );
+    expect(upvoteState, expectedState);
+
+    if (testScore) {
+      final expectedScore = expectedState.increment;
+      expect(await getUpvoteCount(postId), expectedScore);
+    }
+  }
+
+  group(
+    "Operation combination work for a single user with a single post",
+    () {
+      late PostFirestore post;
+      const userId = UserIdFirestore(value: "user_id");
+
+      setUp(() async {
+        final id = await postRepository.addPost(postData, postLocation);
+        post = await postRepository.getPost(id);
+      });
+
+      void testActions(List<UpvoteState> states) {
+        final stateNames = states.map((state) => state.name).join(", ");
+        test(
+          "User can use actions [$stateNames] correctly",
+          () async {
+            await assertPostUpvoteState(userId, post.id, UpvoteState.none);
+
+            for (final state in states) {
+              await postUpvoteRepository.setUpvoteState(
+                userId,
+                post.id,
+                state,
+              );
+
+              await assertPostUpvoteState(userId, post.id, state);
+            }
+          },
+        );
+      }
+
+      for (final state in UpvoteState.values) {
+        testActions([state]);
+      }
+
+      for (final firstState in UpvoteState.values) {
+        for (final secondState in UpvoteState.values) {
+          testActions([firstState, secondState]);
+        }
+      }
+      testActions([
+        UpvoteState.upvoted,
+        UpvoteState.downvoted,
+        UpvoteState.none,
+        UpvoteState.upvoted,
+        UpvoteState.none,
+        UpvoteState.upvoted,
+        UpvoteState.upvoted,
+        UpvoteState.downvoted,
+        UpvoteState.none,
+      ]);
+    },
+  );
+
+  group(
+    "Operations work for a single user on multiple posts",
+    () {
+      const nPosts = 100;
+      late List<PostFirestore> posts;
+      const userId = UserIdFirestore(value: "user_id");
+
+      setUp(() async {
+        posts = [];
+        for (int i = 0; i < nPosts; ++i) {
+          final id = await postRepository.addPost(postData, postLocation);
+          posts.add(await postRepository.getPost(id));
+        }
+      });
+
+      for (final newState in UpvoteState.values) {
+        test(
+          "User can apply action ${newState.name} on $nPosts different posts",
+          () async {
+            for (final post in posts) {
+              assertPostUpvoteState(userId, post.id, UpvoteState.none);
```

Comment:
Indeed! I'm guessing this still worked since it does not have any side effect, but it may have weird non-deterministic implications. Well seen! :)
Fixed in 89c870729ae4119d4d6b7a5e6230c7244bc7f8d0.